// @instantdb/react/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstantReactAbstractDatabase = exports.i = exports.Cursors = exports.InstantReactWebDatabase = exports.init_experimental = exports.init = exports.lookup = exports.tx = exports.id = void 0;
const core_1 = require("@instantdb/core");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return core_1.id; } });
Object.defineProperty(exports, "tx", { enumerable: true, get: function () { return core_1.tx; } });
Object.defineProperty(exports, "lookup", { enumerable: true, get: function () { return core_1.lookup; } });
Object.defineProperty(exports, "i", { enumerable: true, get: function () { return core_1.i; } });
const InstantReactAbstractDatabase_1 = __importDefault(require("./InstantReactAbstractDatabase"));
exports.InstantReactAbstractDatabase = InstantReactAbstractDatabase_1.default;
const InstantReactWebDatabase_1 = __importDefault(require("./InstantReactWebDatabase"));
exports.InstantReactWebDatabase = InstantReactWebDatabase_1.default;
const init_1 = require("./init");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return init_1.init; } });
Object.defineProperty(exports, "init_experimental", { enumerable: true, get: function () { return init_1.init_experimental; } });
const Cursors_1 = require("./Cursors");
Object.defineProperty(exports, "Cursors", { enumerable: true, get: function () { return Cursors_1.Cursors; } });
